patch:  simplify fx injection with `[]func() `group:"cancels"``

denopink opened the pull request:
implements #109 

- Implicitly cast `CancelFunc` to `func()`.
- patch `func (ws *Websocket) AddMessageListener`

denopink committed:
patch: simplify fx life cycle

denopink committed:
Merge branch 'main' into denopink/patch/simplify-fx-life-cycle

schmidtw commented on internal/websocket/options.go:
```diff
@@ -273,45 +273,45 @@ func MaxMessageBytes(bytes int64) Option {
 // AddMessageListener adds a message listener to the WS connection.
 // The listener will be called for every message received from the WS.
-func AddMessageListener(listener event.MsgListener, cancel ...*event.CancelFunc) Option {
+func AddMessageListener(listener event.MsgListener, cancel ...*func()) Option {
```
I don't think this is the right way to achieve this.  Can we not explicitly cast the result to a `func()` afterwards?

denopink replied:
for sure! actually we don't need too since it'll be implicitly casted.

replacing our `CancelFunc` with `func()` was inspired by your eventor update https://github.com/xmidt-org/eventor/pull/7/files

schmidtw replied:
Ok, I see the trend is to be more consistent.  Great!

denopink committed:
Merge branch 'main' into denopink/patch/simplify-fx-life-cycle

denopink committed:
chore: update based on pr feedback

- revert the change inspired by the eventor update  https://github.com/xmidt-org/eventor/pull/7/files (replacing our `CancelFunc`s with `func()`)
- instead, implicitly cast CancelFunc to func()

denopink committed:
chore: update based on pr feedback

- return cancel funcs as groups of func lists named `cancels`
- flatten the fx groups `cancels`
- patch `func (ws *Websocket) AddMessageListener`

denopink commented on internal/websocket/ws.go:
```diff
@@ -187,7 +187,7 @@ func (ws *Websocket) HandleWrp(m wrp.Message) error {
 
 // AddMessageListener adds a message listener to the WS connection.
 // The listener will be called for every message received from the WS.
-func (ws *Websocket) AddMessageListener(listener event.MsgListener, cancel ...*event.CancelFunc) event.CancelFunc {
```
I meant to remove `cancel ...*event.CancelFunc` when we integrate the wrphandlers
